sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,493 @@
+        {
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 0)
+            {
+                StopTimer();
+                await FlushAsync(async, PublisherCancellationToken).ConfigureAwait(false);
+                _channel?.Writer.TryComplete();
+                if (async)
+                {
+                    await _readerLoop.ConfigureAwait(false);
+                }
+                else
+                {
+                    #pragma warning disable AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                    _readerLoop.EnsureCompleted();
+                    #pragma warning restore AZC0104 // Use EnsureCompleted() directly on asynchronous method return value.
+                }
+                _flushBarrier.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Ensure nobody's trying to use this after it's been disposed.
+        /// </summary>
+        private void EnsureNotDisposed()
+        {
+            if (_disposed > 0)
+            {
+                throw new ObjectDisposedException(nameof(SearchIndexingBufferedSender<T>));
+            }
+        }
+        #endregion Dispose
+
+        #region Publishing
+        /// <summary>
+        /// Check if we have a full batch ready to send.
+        /// </summary>
+        /// <returns>If we have a full batch ready to send.</returns>
+        private bool HasFullBatch() => IndexingActionsCount >= BatchActionSize;
+
+        /// <summary>
+        /// Publish as many batches are ready.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task that represents the operation.</returns>
+        private async Task PublishAsync(CancellationToken cancellationToken)
+        {
+            // There's no need to let the timer keep running since we're
+            // already submitting
+            StopTimer();
+
+            do
+            {
+                // Prefer pulling from the _retry queue first
+                List<PublisherAction<T>> batch = new List<PublisherAction<T>>();
+                _ = FillBatchFromQueue(batch, _retry) && FillBatchFromQueue(batch, _pending);
+
+                // Submit the batch
+                await SubmitBatchAsync(
+                    batch,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+
+            // Keep going if we have more full batches ready to submit
+            } while (HasFullBatch());
+
+            // Fill as much of the batch as possible from the given queue.
+            // Returns whether the batch is now full.
+            bool FillBatchFromQueue(List<PublisherAction<T>> batch, Queue< PublisherAction<T>> queue)
+            {
+                // TODO: Consider tracking the keys in the batch and requiring them
+                // to be unique to avoid error alignment problems
+
+                while (queue.Count > 0)
```

Comment:
Given how big the queues could be, it may be worth pre-allocating a list with enough capacity to handle at least one of the queues. The `List<T>` will double the capacity, so checking both queues may not be necessary but couldn't hurt. Otherwise, there's going to be a lot of re-allocation and copying.

Reply:
Great suggestion.  I should just make it the final Min(Pending, BatchActionSize) and it'll always be a single allocation.